src/App.tsx

```diff
@@ -1,24 +1,276 @@
+
+    await Promise.all(completedTodosIds.map(id => handlerDeleteTodo(id)));
+  }, []);
+
+  const handlerTodoToggle = useCallback(async (
+    id: number,
+    completed: boolean,
+  ) => {
+    try {
+      setLoadingTodosIds(prevLoaderIds => ([
+        ...prevLoaderIds,
+        id,
+      ]));
+
+      await changeTodo(id, { completed: !completed });
+
+      setTodos(prevTodos => prevTodos.map(todo => {
+        if (todo.id === id) {
+          return { ...todo, completed: !completed };
+        }
+
+        return todo;
+      }));
+    } catch {
+      handleError(ErrorsType.UPDATE);
+    } finally {
+      setLoadingTodosIds(prevLoaderIds => (
+        prevLoaderIds.filter(prevLoaderId => id !== prevLoaderId)
+      ));
+    }
+  }, []);
+
+  const handlerToggleAll = useCallback(async () => {
+    const areAllCompleted = todos.every(todo => todo.completed);
+    const todosIds = areAllCompleted
+      ? todos.map(todo => todo.id)
+      : todos.filter(todo => !todo.completed).map(todo => todo.id);
+
+    await Promise.all(
+      todosIds.map(id => handlerTodoToggle(id, areAllCompleted)),
+    );
+  }, []);
+
+  const handlerChangeTitle = useCallback(async (
+    newTitle: string,
+    todoId: number,
+  ) => {
+    const title = newTitle.trim();
+
+    if (!title) {
+      await handlerDeleteTodo(todoId);
+
+      return;
+    }
+
+    try {
+      setLoadingTodosIds(prevLoadIds => ([
+        ...prevLoadIds,
+        todoId,
+      ]));
+
+      await changeTodo(todoId, { title });
+
+      setTodos(prevTodos => prevTodos.map(todo => {
+        if (todo.id === todoId) {
+          return { ...todo, title: newTitle };
+        }
+
+        return todo;
+      }));
+    } catch {
+      handleError(ErrorsType.UPDATE);
+    } finally {
+      setLoadingTodosIds(prevLoaderIds => (
+        prevLoaderIds.filter(prevLoaderId => todoId !== prevLoaderId)
+      ));
+    }
+  }, []);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">React Todo App - Add and Delete</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <header className="todoapp__header">
+          <HeaderTodo
+            todos={todos}
+            onAddTodo={addNewTodo}
+            onToogleAll={handlerToggleAll}
+          />
+        </header>
+
+        <TodoList
+          todos={visibleTodos}
+          tempTodo={tempTodo}
+          onDelete={handlerDeleteTodo}
+          idsForLoader={loadingTodosIds}
+          onCompleted={handlerTodoToggle}
+          onChangeTitle={handlerChangeTitle}
+        />
+
+        {todos[0] && (
```

Comment:
but why not checking length?

Reply:
Because we'll show 0 to users instead of TodoFooter

Reply:
But we can wrap it to Boolean constructor or compare with 0